exploration/0477-quoted-literals.md

```diff
@@ -0,0 +1,172 @@
+
+  > ```
+  > {{+button title=|Click here!|}Submit{-button}}
+  > ```
+  >
+  > Note that quoted literals cannot contain placeholders, making interpolating data into them impossible.
+  >
+  > ```
+  > -- This is impossible in MessageFormat 2.0.
+  > {{+button title=|Goodbye, {$userName}!|}Sign out{-button}}
+  > ```
+
+- Selector function implementers may want to support exotic characters in variant keys to effectively create "mini-DSLs" for the matching logic:
+
+  > ```
+  > match {$count :myNumber}
+  > when |<10| {A handful.}
+  > when * {Lots.}
+  > ```
+
+- Message authors may want to protect untranslatable strings:
+
+  > ```
+  > {Visit {|http://www.example.com| @translate=false}.}
+  > ```
+  >
+  > See [design proposal 0002](https://github.com/unicode-org/message-format-wg/blob/main/exploration/0002-expression-attributes.md).
+
+- Message authors may want to decorate substrings as being written in a particular language, different from the message's language, for the purpose of accessibility, text-to-speech, and semantic correctness.
+
+  > ```
+  > {The official native name of the Republic of Poland is {|Rzeczpospolita Polska| @lang=pl}.}
+  > ```
+  >
+  > See [design proposal 0002](https://github.com/unicode-org/message-format-wg/blob/main/exploration/0002-expression-attributes.md).
+
+- Developers may want to embed messages with quoted literals in code written in another programming language which uses single or double quotes to delimit strings.
+
+  > ```js
+  > let message = new MessageFormat("en", "{A message with {|a literal|}.}");
+  > ```
+
+- Developers and localization engineers may want to embed messages with quoted literals in a container format, such as JSON.
+
+  > ```json
+  > {
+  >   "msg": "{A message with {|a literal|}.}"
+  > }
+  > ```
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+- **[r1; high priority]** Minimize the need to escape characters inside literals. In particular, choose a delimiter that isn't frequently used in translation content. Having to escape characters inside literals is inconvenient and error-prone when done by hand, and it also introduces the backslash into the message, `\`, which is the escape introducer. The backslash then needs to be escaped too, when the message is embedded in code or containers. (This is how some syntaxes produce the gnarly `\\\`.)
+- **[r2; high priority]** Minimize the need to escape characters when embedding messages in code or containers. In particular, choose a delimiter that isn't frequently used as a string delimiter in programming languages and container formats. However, note that many programming languages also provide alternative ways of delimiting strings, e.g. _raw strings_ or triple-quoted literals.
+- **[r3; high priority]** Minimize the need to change the message in other ways than to escape some of its characters (e.g. rephrase content or change syntax).
```

Comment:
I meant to include here the need to change straight quotes to curly quotes, or to switch from single quotes to double quote, or the other way around.

Reply:
I'm OK with dropping this to medium priority.

Reply:
I would appreciate this priority being dropped a peg, and including a bit more explanation in the text. All the other requirements include a bit of reasoning and commentary to explain themselves.

Reply:
Thank you for the update. This is still missing the sort of rationalisation that is provided for the other requirements. It is not immediately obvious (at least to me) _why_ this is a requirement. What sort of operation on a message source hits this issue?

Reply:
OK, I see that this requirement isn't well worded. I'll try to rephrase it (possibly tomorrow).

As a discussion point for right now, in your commit on dual quoting, you wrote:

> If the container format does not itself support dual quoting, the embedded message's quotes may be adjusted to avoid their escaping.

This is an example of "needing to change the message in other ways than to escape some of its characters" that I wanted to capture here.

Ideally, it should be possible to drop the message verbatim into code or a container. The next best thing (in my mind) is to only escape some characters, do it rarely, and also do it without thinking about it too much. OTOH, H\having to modify the body in any other way, e.g. switch from straight quotes to curly quotes in translation content, or switch between single and double quotes around literals, is more effort (again, in my mind). I think we should prioritize solutions that don't require such edits, but I also acknowledge that this isn't as high on the priority list as r1 and r2.

Reply:
It might be useful to identify a few specific exemplars of likely container formats in which MF2 messages are expected to often be embedded by developers, such as Java, JS, .properties, YAML, JSON, our own resource format. Are there others we should explicitly consider?